presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/rtree/Flatbush.java

```diff
@@ -0,0 +1,306 @@
+     * rectangle intersects the query rectangle.  The actual intersection
+     * check will need to be performed by the caller.
+     *
+     * @param query Rectangle for which to search for intersection.
+     * @param consumer Function to call for each intersection candidate.
+     */
+    public void findIntersections(Rectangle query, Consumer<T> consumer)
+    {
+        IntArrayList todoNodes = new IntArrayList(levelOffsets.length * degree);
+        IntArrayList todoLevels = new IntArrayList(levelOffsets.length * degree);
+
+        int rootLevel = levelOffsets.length - 1;
+        int rootIndex = levelOffsets[rootLevel];
+        if (doesIntersect(query, rootIndex)) {
+            todoNodes.push(rootIndex);
+            todoLevels.push(rootLevel);
+        }
+
+        while (!todoNodes.isEmpty()) {
+            int nodeIndex = todoNodes.popInt();
+            int level = todoLevels.popInt();
+
+            if (level == 0) {
+                // This is a leaf node
+                consumer.accept(items[nodeIndex / ENVELOPE_SIZE]);
+            }
+            else {
+                int childrenOffset = getChildrenOffset(nodeIndex, level);
+                for (int i = 0; i < degree; i++) {
+                    int childIndex = childrenOffset + ENVELOPE_SIZE * i;
+                    if (doesIntersect(query, childIndex)) {
+                        todoNodes.push(childIndex);
+                        todoLevels.push(level - 1);
+                    }
+                }
+            }
+        }
+    }
+
+    private boolean doesIntersect(Rectangle query, int nodeIndex)
+    {
+        return query.getXMax() >= tree.get(nodeIndex) // xMin
+                && query.getYMax() >= tree.get(nodeIndex + 1) // yMin
+                && query.getXMin() <= tree.get(nodeIndex + 2) // xMax
+                && query.getYMin() <= tree.get(nodeIndex + 3); // yMax
+    }
+
+    /**
+     * Get the offset of the first child for the node.
+     *
+     * @param nodeIndex Index in tree of first entry for node
+     * @param level Level of node
+     */
+    @VisibleForTesting
+    int getChildrenOffset(int nodeIndex, int level)
+    {
+        int indexInLevel = nodeIndex - levelOffsets[level];
+        return levelOffsets[level - 1] + degree * indexInLevel;
+    }
+
+    @VisibleForTesting
+    int getHeight()
+    {
+        return levelOffsets.length;
+    }
+
+    /*
+     * Sorts items in-place by the Hilbert index of the envelope center.
+     */
+    private void sortByHilbertIndex(T[] items)
+    {
+        if (items == null || items.length < 2) {
+            return;
+        }
+
+        Rectangle totalExtent = items[0].getExtent();
+        for (int i = 1; i < items.length; i++) {
+            totalExtent = totalExtent.merge(items[i].getExtent());
+        }
+
+        HilbertIndex hilbert = new HilbertIndex(totalExtent);
+        Arrays.parallelSort(items, Comparator.comparing(item ->
```

Comment:
@jagill is it inefficient to compute the hilbert index dynamically like this?  Won't there be many computations of the same index?

Reply:
The index is calculated on average `O(ln N)` times per entry, as you said.  I originally used davidmoten/hilbert-curve, and (in the profiling) index calculation did indeed dominate construction CPU.  However, this new calculation is very efficient (all bit twiddling), and so now it's only a few % of construction CPU.

Reply:
Ok, makes sense, and certainly simpler than maintaining a pre-computed value.